Pal 874 data service update documentation (#277)
gchq/Palisade-services: data-service/README.md

--->
# Data Service

> The Data Service accepts client requests to retrieve resources that have been registered on their behalf. These will be

Reviewer: Linebreaks across md file

Author: fixed

the response to the initial requests sent to the Palisade service that have been collected, filtered and possibly
redacted in conformance to the defined rules and to the context of the request. The client is expected to send a
request containing the token and resource id that is used to uniquely identify the resource request. The response will
be an output stream holding the data resources. To see more information on client requests, see the Palisade Client's
library.

The key components of the service is the implementation of the DataService interface and the supporting services.  This
comes in the form of an implementation of DataService interface, SimpleDataService, that uses the Palisade Reader
library for the implementation of a solution using a database. This is then wrapped in a class, AuditableDataService,

> that will provide the data in a form that can be use in constructing the response that is sent back to the client, and

Reviewer 2: ```suggestion
that will provide the data in a form that can be used in constructing the response that is sent back to the client, and
```

the audit message that is sent to the Audit Service using the AuditMessageService.

```java
interface DataService extends Service { {
  CompletableFuture<AuthorisedDataRequest> authoriseRequest(final DataRequest request);

  CompletableFuture<Boolean> read(final AuthorisedDataRequest request, final OutputStream out, final AtomicLong recordsProcessed, final AtomicLong recordsReturned);

}

> ```

Reviewer: Link class rather than copy code snippet

Author: done

# Flow of Control

> o![Data Service diagram](doc/data-service.png)

Reviewer 2: ```suggestion
![Data Service diagram](doc/data-service.png)
```
The flow image should be colour coordinated with kafka sources being purple, for example. A good example of the colours in use can be seen in the [FRS flow diagram](https://github.com/gchq/Palisade-services/blob/develop/filtered-resource-service/doc/filtered-resource-service.png).
The lines in the diagram are not easy to see so it might be worth while looking at what I've done for the [User Service](https://github.com/gchq/Palisade-services/blob/Pal-872-user-service-streaming-documentation/user-service/doc/user-service.png)

> ![Data Service diagram](doc/data-service.png)

Reviewer: "Request Authoriseded" needs correcting
Also, an arrow to a db to match with attr-mask-svc?

Author: Added a db for the authorise request and the read in the diagram

> Data Service is implemented as a RESTFul Request-Response service using a Spring Controller as the service endpoint.  

Reviewer: ```suggestion
The Data Service is implemented as a RESTFul service using a Spring Controller as the service endpoint.  
```

Author: done

> The processing of each client request will drive the workflow though the service. This will first start with the 

Reviewer 2: ```suggestion
The processing of each client request will drive the workflow though the service. This will start with 
```

authorising the request (`DataService#authoriseReqeust`). If successful, the next step will be to read the data

> (`DataService#read`) which will return a stream of the data to the client. Upon completion, a audit message is sent to 

Reviewer 2: ```suggestion
(`DataService#read`) which will return a stream of the data to the client. Upon completion, an audit message is sent to 
```

the Audit Service indicating a successful transfer of the data. If during the authorisation request or at any time in 
the read request there is a error, the request is stopped, and an error message is sent to the Audit Service.

Kafka streaming is only used for the processing of audit messages. This is exposed in the functionality provided by the

> `AuditMessageService`.  This will take both audit success and error messages and process them using the functionality 

Reviewer 2: ```suggestion
`AuditMessageService`.  This will take both audit success and audit error messages and process them using the functionality 
```

|                 |                        | serverMetadata       | error             |
|                 |                        |                      | serverMetadata    |
  
*token comes in the body of the request from the client (DataRequest) and is stored in the header metadata for the 
audit messages

**attributes will include the numbers for records processed and records returned

> ***data types that may not be available depending on when the error occurred

Reviewer: ```suggestion
***data that may not be available depending on when the error occurred
```
The types are always available, the data not so

## REST Interface

The application exposes one endpoint to the client for retrieving the resources. This will be the data that has previously 

> been requested and prepared in the initial request to Palisade services. 

Reviewer 2: ```suggestion
been requested and prepared in the initial request to the Palisade services. 
```

* `POST data/read/chunked`

>     - returns a `200 OK` and a streamed HTTP response body which will provides the resource.

Reviewer 2: ```suggestion
    - returns a `200 OK` and a streamed HTTP response body which will provide the resource(s).
```

>     - returns a `200 OK` and a streamed HTTP response body which will provide the resource(s).

Reviewer: Having changed these before, and seeing that some things actually care about the response, do you want to make this a nondescript `2xx`? Not an issue, just a thought

Author: The short of it is that I don't know.  A 202 tells us that they "got" the message, but we don't know what they did with it.  The successful response we do get back will be one with an output stream and no guarantee that everything will work.  What kind of a 200 status is that?

Reviewer: Getting into the nitty-gritty - 
https://tools.ietf.org/html/rfc7231#section-6.3.1
Since we always return a body (even if it's a stream), it should be a 200 OK, so no change needed
